tests/suites/test_suite_mpi.function

```diff
@@ -1720,6 +1720,378 @@ exit:
 }
 /* END_CASE */
 
+/* BEGIN_CASE */
+void mbedtls_mpi_core_add_if( char * input_l, char * input_r,
+                              char * input_X4, int carry4,
+                              char * input_X8, int carry8 )
```

Comment:
I find both the test descriptions and the test code hard to read. Please switch to having two separate test cases when the result is different depending on the limb size, each conditioned on a particular limb size. This would both make the code and test cases more readable, and make it easier to spot limb-size-specific failures since they'd appear in different test cases.

I'm not sure if we have a preprocessor condition for the mpi limb size. It's determined by some fairly complex logic in `include/mbedtls/bignum.h` and I suggest adding `#define MBEDTLS_MPI_LIMB_SIZE 4` or `#define MBEDTLS_MPI_LIMB_SIZE 8` (or `_BITS` set to 32 or 64) to each case.

Reply:
The two constants are already present, `MBEDTLS_HAVE_INT32` and `MBEDTLS_HAVE_INT64`. I already use those for cases where it wasn't practicable to use the same test cases for different limb size.

Again, our minds work completely differently.

To me, a test case for `add_if` (for example) might be `0x1 + 0xffffffff`. Now, this gives different results depending on whether the limb size is 4 or 8, but fundamentally it's the same test case. It's a simple bit of boilerplate code to select which of X4 or X8 (for example) are needed in the test code, and assign it to X (which is all that is used subsequently in the test functions).

Making such a change is going to require me to rewrite all my test scripts and rewrite the test functions.

Is the benefit of this actually worth the cost?

Reply:
I had a hard time reading the test code. It seemed like more than half the code was to cater for word size differences or encoding woes between the old and new representations. I'm willing to give it another go, but no promises that I'll come to a different conclusion.

I don't consider 0x1 + 0xffffffff to be a useful test case. What _is_ a useful test case is 1 + LIMB_MAX, testing a simple case with a carry. LIMB_MAX depends on the limb size and there's no really convenient way to spell that in the test data. It could be expressed by having different sets of _inputs_ for the same test case depending on the limb size, with a common expected carry value. That way there's a single test case for the same semantics, but at the cost of making the test data and the test code more complex. Having different test cases would make the test code significantly simpler, and the test data not much more complex. If the test data was generated from a script, which is a direction I'd like to move towards, the test data complexity would be pretty much smoothed over. As it is now, having different sets of outputs is the worst of both worlds: more complex test code and data, _and_ no clear semantics for the test case.

Reply:
> If the test data was generated from a script

This test data WAS all generated by a script

Reply:
Ah, my bad, I'd stopped trying to puzzle out the code and didn't look at the commit message. (f334d9622b4c327c56ff081c39fd4b5cb3a92753)

The 4/8 duality practically doubles the script size... It's making everything more complex, with no redeeming feature that I can see.

For future reference, we're moving to having this kind of scripts in the code base, and we're using Python for all scripts of this kind. We old farts know Perl, but young'uns don't and I don't see why they should learn.

Reply:
> we're moving to having this kind of scripts in the code base, and we're using Python for all scripts of this kind

Yes, I know. However, Werner's stuff wasn't ready when I started the bignum work, I don't know Python at all well, and I didn't want to add "get familiar enough with Python" to all the things I was doing while working on this PR (I used the Perl script to play around and get familiar with bignum stuff and montmul, and doing that in a familiar language was pretty much essential).

> We old farts know Perl, but young'uns don't and I don't see why they should learn

Unfortunately I agree :)

Reply:
For this specific application, Python has the advantage that it has bignums built in, none of this bothering with `Math::BigInt` conversions and explicit copy and stuff. Modular exponentiation is also built in (`pow(a, b, p)` = a^b mod p), but not inverse (actually inverse got added in Python 3.8 — yay — but we still support down to 3.5).